### ISSUE TITLE
dropout does not take in dynamic shape

### ISSUE THREAD

**Issue by hycis** [tensorflow/tensorflow#7224]
I realize dropout `noise_shape` does not allow dynamic shape which becomes useless in batch processing whereby the batchsize can varies
```python
X = tf.placeholder('float', [None, 5, 10])
tf.dropout(X, keep_prob=0.5, noise_shape=[-1, 5, 1]) 
```
throws Exception. Whereby it requires dimension to be >= 0

however if we pass noise_shape as the shape of X which is dynamic, dropout can understand the shape.
```python
tf.dropout(X, keep_prob=0.5, noise_shape=tf.shape(X))
```
Therefore there is inconsistency in noise_shape inputs. How do we give dynamic shapes to `noise_shape` such as `noise_shape=[-1, 5, 1]` where the batchsize dimension can be variable just like how `tf.reshape` works

```python
tf.reshape(X, [-1, 5, 1])
```

**Comment by girving** [comment]
There are only a few functions that understand negative arguments to shape, such as `tf.reshape`.  Generalizing every function that takes a shape to support them wherever possible would overly complicate the code (currently, `tf.dropout` just passes the shape to `tf.random_uniform`.  Since dropout already does support dynamic shape, as you note, I'll close this.

**Comment by hycis** [comment]
@girving So in this how do we create a dropout mask for batch training of variable batch size, for example in a sequence of text embeddings of dimension `(batchsize, seqlen, embed_dim)` where we want to randomly dropout the some of the word embedding in the sequence, which will require noise shape of `noise_shape=[-1, seqlen, 1]`, then how we do it with the current dropout?

**Comment by girving** [comment]
`noise_shape=[tf.shape(...)[0], seqlen, 1]`

**Comment by hycis** [comment]
Thanks,  I can make the noise_shape able to take in -1 by simply converting a -1 at i position to a tf.shape(X)[i], i can make a pull request if that works.

**Comment by girving** [comment]
@hycis There are at least two problems with your suggestion already: you're off if the noise shape is lower rank, and your solution can't handle inputs that are already tensors.  There's also the larger question that I don't think this trick is standard numpy, and I'm leery of settling on an API that isn't standardized.
